subprojects/architecture-test/src/test/java/org/gradle/architecture/test/KotlinCompatibilityTest.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.architecture.test;
+
+import com.tngtech.archunit.core.domain.JavaClass;
+import com.tngtech.archunit.core.domain.JavaMethod;
+import com.tngtech.archunit.junit.AnalyzeClasses;
+import com.tngtech.archunit.junit.ArchTest;
+import com.tngtech.archunit.junit.ArchUnitRunner;
+import com.tngtech.archunit.lang.ArchCondition;
+import com.tngtech.archunit.lang.ArchRule;
+import com.tngtech.archunit.lang.ConditionEvents;
+import org.junit.runner.RunWith;
+
+import java.util.Arrays;
+import java.util.Set;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+
+import static com.tngtech.archunit.lang.SimpleConditionEvent.violated;
+import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.classes;
+import static java.util.stream.Collectors.toSet;
+
+@RunWith(ArchUnitRunner.class)
+@AnalyzeClasses(packages = "org.gradle")
+public class KotlinCompatibilityTest {
+
+    @ArchTest
+    public static final ArchRule consistent_nullable_annotations = classes().should(haveAccessorsWithSymmetricalNullableAnnotations());
+
+    private static ArchCondition<JavaClass> haveAccessorsWithSymmetricalNullableAnnotations() {
+        return new ArchCondition<JavaClass>("have accessors with symmetrical @Nullable annotations") {
+            @Override
+            public void check(JavaClass item, ConditionEvents events) {
+                findSymmetricalAccessors(item).stream()
+                    .filter(this::nullableIsUnsymmetrical)
+                    .map(accessor -> violated(item, String.format("Accessors for %s don't use symmetrical @Nullable", accessor)))
+                    .forEach(events::add);
+            }
+
+            private Set<Accessors> findSymmetricalAccessors(JavaClass javaClass) {
+                return filterGetters(javaClass.getMethods()).stream()
+                    .flatMap(getter -> tryFindMatchingSetters(javaClass, getter).map(s -> new Accessors(getter, s)))
+                    .collect(toSet());
+            }
+
+            private Set<JavaMethod> filterGetters(Set<JavaMethod> methods) {
+                return methods.stream().filter(m -> m.getName().startsWith("get") && m.getName().length() > 3).collect(toSet());
```

Comment:
:x: this won't find boolean `isName` getters. We should try reuse what we do to e.g. find properties on task types in order to handle all JavaBean spec corner cases.

Reply:
I am re-using some of the infrastructure now. I am not very happy about the usage, since I cannot use it fully without adding some extra dependencies to the architecture tests. That is due to the fact that we have some `compileOnly` dependencies like e.g. Zinc and cannot load the classes if those are not on the classpath.

Do you think it is good as it is now or shall I add the dependencies to the classpath?

Reply:
Got it. It's already much better like that 👍